auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditOrganisationsController.kt

```diff
@@ -0,0 +1,108 @@
+package uk.gov.communities.delta.auth.controllers.internal
+
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.auth.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import kotlinx.serialization.Serializable
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.config.DeltaConfig
+import uk.gov.communities.delta.auth.plugins.ApiError
+import uk.gov.communities.delta.auth.repositories.LdapUser
+import uk.gov.communities.delta.auth.services.*
+
+class EditOrganisationsController(
+    private val userLookupService: UserLookupService,
+    private val groupService: GroupService,
+    private val organisationService: OrganisationService,
+    private val accessGroupsService: AccessGroupsService,
+    private val memberOfToDeltaRolesMapperFactory: MemberOfToDeltaRolesMapperFactory,
+) {
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    fun route(route: Route) {
+        route.post { updateUserOrganisations(call) }
+    }
+
+    // This endpoint takes a list of organisation codes which should contain all and only the organisations of which
+    // the user should be a member after the execution of the endpoint
+    private suspend fun updateUserOrganisations(call: ApplicationCall) {
+        val session = call.principal<OAuthSession>()!!
+        val callingUser = userLookupService.lookupUserByCn(session.userCn)
+        logger.atInfo().log("Updating organisations for user {}", session.userCn)
+
+        val requestedOrganisations = call.receive<DeltaUserOrganisations>().userSelectedOrgs
+        val userDomainOrgs =
+            if (callingUser.email == null) mapOf() else organisationService.findAllByDomain(callingUser.email)
+                .associateBy { it.code }
+        validateOrganisationRequest(requestedOrganisations, userDomainOrgs)
+
+        val existingOrganisations = getExistingOrganisationsForUser(callingUser)
+
+        val orgsToAdd = requestedOrganisations.toSet().minus(existingOrganisations.toSet())
+        val orgsToRemove = existingOrganisations.toSet().minus(requestedOrganisations.toSet())
```

Comment:
We shouldn't remove the user from non-domain organisations, admins can assign users extra organisations that don't match there domain.

We could:
* Just ignore non-domain orgs here
* Have separate organisationsToAdd and organisationsToRemove passed in
* Have separate endpoints that add or remove one organisation

To be honest I think the last option might be nicest? Given the number of organisations is normally very small. Up to you though, if we're just ignoring them we should make it very clear that's what this endpoint does.